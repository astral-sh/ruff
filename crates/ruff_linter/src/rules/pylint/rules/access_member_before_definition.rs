use ruff_diagnostics::{Diagnostic, FixAvailability, Violation};
use ruff_macros::{derive_message_formats, violation};
use ruff_python_semantic::{ScopeId, ScopeKind};

use ruff_source_file::SourceRow;
use ruff_text_size::Ranged;

use crate::checkers::ast::Checker;

/// ## What it does
/// Checks for accesses on instance attributes before they are initialized in __init__ method.
/// If the attribute is not initialized at all then it is not reported.
///
/// ## Why is this bad?
/// Accessing a class member before it is initialized will raise an `AttributeError`.
///
/// ## Example
/// ```python
/// class Unicorn:
///     def __init__(self, x):
///         if self.x > 9000:  # [access-member-before-definition]
///             pass
///         self.x = x
/// ```
///
/// Use instead:
/// ```python
/// class Unicorn:
///     def __init__(self, x):
///         self.x = x
///         if self.x > 9000:
///             pass
/// ```
///
#[violation]
pub struct AccessMemberBeforeDefinition {
    member: String,
    initialized_at: SourceRow,
}

impl Violation for AccessMemberBeforeDefinition {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::None;

    #[derive_message_formats]
    fn message(&self) -> String {
        let AccessMemberBeforeDefinition {
            member,
            initialized_at,
        } = self;
        format!("Accessed `{member}` before it was initialized at {initialized_at}",)
    }
}

/// PLE0203
pub(crate) fn access_member_before_definition(
    checker: &Checker,
    scope_id: ScopeId,
    diagnostics: &mut Vec<Diagnostic>,
) {
    if let Some(scope) = checker.semantic().scopes.get(scope_id) {
        // Only run on __init__ method
        match scope.kind {
            ScopeKind::Function(function) => {
                if function.name.as_str() != "__init__" {
                    return;
                }
            }
            _ => return,
        };
    };

    let Some(enclosing_class_scope_id) =
        checker.semantic().first_non_type_parent_scope_id(scope_id)
    else {
        return;
    };

    let Some(enclosing_class_scope) = checker.semantic().scopes.get(enclosing_class_scope_id)
    else {
        return;
    };

    for unresolved in checker.semantic().unresolved_attributes() {
        if unresolved.scope_id != scope_id {
            continue;
        }
        let attr_name = checker.locator().slice(unresolved.range);
        if let Some(binding_id) = enclosing_class_scope.get(attr_name) {
            let binding = checker.semantic().binding(binding_id);
            if binding.start() > unresolved.range.end() {
                let diagnostic = Diagnostic::new(
                    AccessMemberBeforeDefinition {
                        initialized_at: checker.compute_source_row(binding.start()),
                        member: attr_name.to_string(),
                    },
                    unresolved.range,
                );
                diagnostics.push(diagnostic);
            }
        };
    }
}
