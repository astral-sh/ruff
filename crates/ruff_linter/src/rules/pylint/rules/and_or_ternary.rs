use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
use ruff_macros::{derive_message_formats, violation};
use ruff_python_ast::{
    BoolOp, Expr, ExprBoolOp, ExprDictComp, ExprIfExp, ExprListComp, ExprSetComp,
};
use ruff_text_size::{Ranged, TextRange};

use crate::checkers::ast::Checker;
use crate::fix::snippet::SourceCodeSnippet;
use crate::registry::AsRule;

/// ## What it does
/// Checks for uses of the known pre-Python 2.5 ternary syntax.
///
/// ## Why is this bad?
/// Prior to the introduction of the if-expression (ternary) operator in Python
/// 2.5, the only way to express a conditional expression was to use the `and`
/// and `or` operators.
///
/// The if-expression construct is clearer and more explicit, and should be
/// preferred over the use of `and` and `or` for ternary expressions.
///
/// ## Example
/// ```python
/// x, y = 1, 2
/// maximum = x >= y and x or y
/// ```
///
/// Use instead:
/// ```python
/// x, y = 1, 2
/// maximum = x if x >= y else y
/// ```
#[violation]
pub struct AndOrTernary {
    ternary: SourceCodeSnippet,
}

impl Violation for AndOrTernary {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;

    #[derive_message_formats]
    fn message(&self) -> String {
        if let Some(ternary) = self.ternary.full_display() {
            format!("Consider using if-else expression (`{ternary}`)")
        } else {
            format!("Consider using if-else expression")
        }
    }

    fn fix_title(&self) -> Option<String> {
        Some(format!("Convert to if-else expression"))
    }
}

/// Returns `Some((condition, true_value, false_value))`, if `bool_op` is of the form `condition and true_value or false_value`.
fn parse_and_or_ternary(bool_op: &ExprBoolOp) -> Option<(&Expr, &Expr, &Expr)> {
    if bool_op.op != BoolOp::Or {
        return None;
    }
    let [expr, false_value] = bool_op.values.as_slice() else {
        return None;
    };
    let Some(and_op) = expr.as_bool_op_expr() else {
        return None;
    };
    if and_op.op != BoolOp::And {
        return None;
    }
    let [condition, true_value] = and_op.values.as_slice() else {
        return None;
    };
    if false_value.is_bool_op_expr() || true_value.is_bool_op_expr() {
        return None;
    }
    Some((condition, true_value, false_value))
}

/// Returns `true` if the expression is used within a comprehension.
fn is_comprehension_if(parent: Option<&Expr>, expr: &ExprBoolOp) -> bool {
    let comprehensions = match parent {
        Some(Expr::ListComp(ExprListComp { generators, .. })) => generators,
        Some(Expr::SetComp(ExprSetComp { generators, .. })) => generators,
        Some(Expr::DictComp(ExprDictComp { generators, .. })) => generators,
        _ => {
            return false;
        }
    };
    comprehensions
        .iter()
        .any(|comp| comp.ifs.iter().any(|ifs| ifs.range() == expr.range()))
}

/// PLR1706
pub(crate) fn and_or_ternary(checker: &mut Checker, bool_op: &ExprBoolOp) {
    if checker.semantic().current_statement().is_if_stmt() {
        return;
    }
    let parent_expr = checker.semantic().current_expression_parent();
    if parent_expr.is_some_and(Expr::is_bool_op_expr) {
        return;
    }
    let Some((condition, true_value, false_value)) = parse_and_or_ternary(bool_op) else {
        return;
    };

    let if_expr = Expr::IfExp(ExprIfExp {
        test: Box::new(condition.clone()),
        body: Box::new(true_value.clone()),
        orelse: Box::new(false_value.clone()),
        range: TextRange::default(),
    });

    let ternary = if is_comprehension_if(parent_expr, bool_op) {
        format!("({})", checker.generator().expr(&if_expr))
    } else {
        checker.generator().expr(&if_expr)
    };

    let mut diagnostic = Diagnostic::new(
        AndOrTernary {
            ternary: SourceCodeSnippet::new(ternary.clone()),
        },
        bool_op.range,
    );
    if checker.enabled(diagnostic.kind.rule()) {
        diagnostic.set_fix(Fix::unsafe_edit(Edit::range_replacement(
            ternary,
            bool_op.range,
        )));
    }
    checker.diagnostics.push(diagnostic);
}
