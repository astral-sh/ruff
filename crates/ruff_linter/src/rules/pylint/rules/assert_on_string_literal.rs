use ruff_python_ast::{self as ast, Expr};

use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, violation};
use ruff_text_size::Ranged;

use crate::checkers::ast::Checker;

#[derive(Debug, PartialEq, Eq, Copy, Clone)]
enum Kind {
    Empty,
    NonEmpty,
    Unknown,
}

/// ## What it does
/// Checks for `assert` statements that use a string literal as the first
/// argument.
///
/// ## Why is this bad?
/// An `assert` on a non-empty string literal will always pass, while an
/// `assert` on an empty string literal will always fail.
///
/// ## Example
/// ```python
/// assert "always true"
/// ```
#[violation]
pub struct AssertOnStringLiteral {
    kind: Kind,
}

impl Violation for AssertOnStringLiteral {
    #[derive_message_formats]
    fn message(&self) -> String {
        let AssertOnStringLiteral { kind } = self;
        match kind {
            Kind::Empty => format!("Asserting on an empty string literal will never pass"),
            Kind::NonEmpty => format!("Asserting on a non-empty string literal will always pass"),
            Kind::Unknown => format!("Asserting on a string literal may have unintended results"),
        }
    }
}

/// PLW0129
pub(crate) fn assert_on_string_literal(checker: &mut Checker, test: &Expr) {
    match test {
        Expr::StringLiteral(ast::ExprStringLiteral { value, .. }) => {
            checker.diagnostics.push(Diagnostic::new(
                AssertOnStringLiteral {
                    kind: if value.is_empty() {
                        Kind::Empty
                    } else {
                        Kind::NonEmpty
                    },
                },
                test.range(),
            ));
        }
        Expr::BytesLiteral(ast::ExprBytesLiteral { value, .. }) => {
            checker.diagnostics.push(Diagnostic::new(
                AssertOnStringLiteral {
                    kind: if value.is_empty() {
                        Kind::Empty
                    } else {
                        Kind::NonEmpty
                    },
                },
                test.range(),
            ));
        }
        Expr::FString(ast::ExprFString { value, .. }) => {
            let kind = if value.iter().all(|f_string_part| match f_string_part {
                ast::FStringPart::Literal(literal) => literal.is_empty(),
                ast::FStringPart::FString(f_string) => {
                    f_string.elements.iter().all(|element| match element {
                        ast::FStringElement::Literal(ast::FStringLiteralElement {
                            value, ..
                        }) => value.is_empty(),
                        ast::FStringElement::Expression(_) => false,
                    })
                }
            }) {
                Kind::Empty
            } else if value.iter().any(|f_string_part| match f_string_part {
                ast::FStringPart::Literal(literal) => !literal.is_empty(),
                ast::FStringPart::FString(f_string) => {
                    f_string.elements.iter().any(|element| match element {
                        ast::FStringElement::Literal(ast::FStringLiteralElement {
                            value, ..
                        }) => !value.is_empty(),
                        ast::FStringElement::Expression(_) => false,
                    })
                }
            }) {
                Kind::NonEmpty
            } else {
                Kind::Unknown
            };
            checker.diagnostics.push(Diagnostic::new(
                AssertOnStringLiteral { kind },
                test.range(),
            ));
        }
        _ => {}
    }
}
