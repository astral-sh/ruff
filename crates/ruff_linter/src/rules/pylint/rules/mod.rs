pub(crate) use and_or_ternary::*;
pub(crate) use assert_on_string_literal::*;
pub(crate) use await_outside_async::*;
pub(crate) use bad_dunder_method_name::*;
pub(crate) use bad_str_strip_call::*;
pub(crate) use bad_string_format_character::BadStringFormatCharacter;
pub(crate) use bad_string_format_type::*;
pub(crate) use bidirectional_unicode::*;
pub(crate) use binary_op_exception::*;
pub(crate) use collapsible_else_if::*;
pub(crate) use compare_to_empty_string::*;
pub(crate) use comparison_of_constant::*;
pub(crate) use comparison_with_itself::*;
pub(crate) use continue_in_finally::*;
pub(crate) use duplicate_bases::*;
pub(crate) use eq_without_hash::*;
pub(crate) use global_statement::*;
pub(crate) use global_variable_not_assigned::*;
pub(crate) use import_self::*;
pub(crate) use invalid_all_format::*;
pub(crate) use invalid_all_object::*;
pub(crate) use invalid_envvar_default::*;
pub(crate) use invalid_envvar_value::*;
pub(crate) use invalid_str_return::*;
pub(crate) use invalid_string_characters::*;
pub(crate) use iteration_over_set::*;
pub(crate) use load_before_global_declaration::*;
pub(crate) use logging::*;
pub(crate) use magic_value_comparison::*;
pub(crate) use manual_import_from::*;
pub(crate) use named_expr_without_context::*;
pub(crate) use nested_min_max::*;
pub(crate) use no_self_use::*;
pub(crate) use nonlocal_without_binding::*;
pub(crate) use property_with_parameters::*;
pub(crate) use redefined_loop_name::*;
pub(crate) use repeated_equality_comparison::*;
pub(crate) use repeated_isinstance_calls::*;
pub(crate) use return_in_init::*;
pub(crate) use self_assigning_variable::*;
pub(crate) use single_string_slots::*;
pub(crate) use subprocess_popen_preexec_fn::*;
pub(crate) use subprocess_run_without_check::*;
pub(crate) use sys_exit_alias::*;
pub(crate) use too_many_arguments::*;
pub(crate) use too_many_branches::*;
pub(crate) use too_many_public_methods::*;
pub(crate) use too_many_return_statements::*;
pub(crate) use too_many_statements::*;
pub(crate) use type_bivariance::*;
pub(crate) use type_name_incorrect_variance::*;
pub(crate) use type_param_name_mismatch::*;
pub(crate) use unexpected_special_method_signature::*;
pub(crate) use unnecessary_direct_lambda_call::*;
pub(crate) use useless_else_on_loop::*;
pub(crate) use useless_import_alias::*;
pub(crate) use useless_return::*;
pub(crate) use yield_from_in_async_function::*;
pub(crate) use yield_in_init::*;

mod and_or_ternary;
mod assert_on_string_literal;
mod await_outside_async;
mod bad_dunder_method_name;
mod bad_str_strip_call;
pub(crate) mod bad_string_format_character;
mod bad_string_format_type;
mod bidirectional_unicode;
mod binary_op_exception;
mod collapsible_else_if;
mod compare_to_empty_string;
mod comparison_of_constant;
mod comparison_with_itself;
mod continue_in_finally;
mod duplicate_bases;
mod eq_without_hash;
mod global_statement;
mod global_variable_not_assigned;
mod import_self;
mod invalid_all_format;
mod invalid_all_object;
mod invalid_envvar_default;
mod invalid_envvar_value;
mod invalid_str_return;
mod invalid_string_characters;
mod iteration_over_set;
mod load_before_global_declaration;
mod logging;
mod magic_value_comparison;
mod manual_import_from;
mod named_expr_without_context;
mod nested_min_max;
mod no_self_use;
mod nonlocal_without_binding;
mod property_with_parameters;
mod redefined_loop_name;
mod repeated_equality_comparison;
mod repeated_isinstance_calls;
mod return_in_init;
mod self_assigning_variable;
mod single_string_slots;
mod subprocess_popen_preexec_fn;
mod subprocess_run_without_check;
mod sys_exit_alias;
mod too_many_arguments;
mod too_many_branches;
mod too_many_public_methods;
mod too_many_return_statements;
mod too_many_statements;
mod type_bivariance;
mod type_name_incorrect_variance;
mod type_param_name_mismatch;
mod unexpected_special_method_signature;
mod unnecessary_direct_lambda_call;
mod useless_else_on_loop;
mod useless_import_alias;
mod useless_return;
mod yield_from_in_async_function;
mod yield_in_init;
