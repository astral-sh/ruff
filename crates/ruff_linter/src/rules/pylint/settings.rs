//! Settings for the `pylint` plugin.

use rustc_hash::FxHashSet;
use serde::{Deserialize, Serialize};

use ruff_macros::CacheKey;
use ruff_python_ast::{ExprNumberLiteral, LiteralExpressionRef, Number};

#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize, CacheKey)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
pub enum ConstantType {
    Bytes,
    Complex,
    Float,
    Int,
    Str,
}

impl ConstantType {
    pub fn try_from_literal_expr(literal_expr: LiteralExpressionRef<'_>) -> Option<Self> {
        match literal_expr {
            LiteralExpressionRef::StringLiteral(_) => Some(Self::Str),
            LiteralExpressionRef::BytesLiteral(_) => Some(Self::Bytes),
            LiteralExpressionRef::NumberLiteral(ExprNumberLiteral { value, .. }) => match value {
                Number::Int(_) => Some(Self::Int),
                Number::Float(_) => Some(Self::Float),
                Number::Complex { .. } => Some(Self::Complex),
            },
            LiteralExpressionRef::BooleanLiteral(_)
            | LiteralExpressionRef::NoneLiteral(_)
            | LiteralExpressionRef::EllipsisLiteral(_) => None,
        }
    }
}

#[derive(Debug, CacheKey)]
pub struct Settings {
    pub allow_magic_value_types: Vec<ConstantType>,
    pub allow_dunder_method_names: FxHashSet<String>,
    pub max_args: usize,
    pub max_positional_args: usize,
    pub max_returns: usize,
    pub max_bool_expr: usize,
    pub max_branches: usize,
    pub max_statements: usize,
    pub max_public_methods: usize,
    pub max_locals: usize,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            allow_magic_value_types: vec![ConstantType::Str, ConstantType::Bytes],
            allow_dunder_method_names: FxHashSet::default(),
            max_args: 5,
            max_positional_args: 5,
            max_returns: 6,
            max_bool_expr: 5,
            max_branches: 12,
            max_statements: 50,
            max_public_methods: 20,
            max_locals: 15,
        }
    }
}
