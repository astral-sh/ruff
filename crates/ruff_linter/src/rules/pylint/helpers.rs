use std::fmt;

use ruff_python_ast as ast;
use ruff_python_ast::{Arguments, CmpOp, Constant, Expr};
use ruff_python_semantic::analyze::function_type;
use ruff_python_semantic::{ScopeKind, SemanticModel};

use crate::settings::LinterSettings;

/// Returns the value of the `name` parameter to, e.g., a `TypeVar` constructor.
pub(super) fn type_param_name(arguments: &Arguments) -> Option<&str> {
    // Handle both `TypeVar("T")` and `TypeVar(name="T")`.
    let name_param = arguments.find_argument("name", 0)?;
    if let Expr::Constant(ast::ExprConstant {
        value: Constant::Str(name),
        ..
    }) = &name_param
    {
        Some(name)
    } else {
        None
    }
}

pub(super) fn in_dunder_init(semantic: &SemanticModel, settings: &LinterSettings) -> bool {
    let scope = semantic.current_scope();
    let ScopeKind::Function(ast::StmtFunctionDef {
        name,
        decorator_list,
        ..
    }) = scope.kind
    else {
        return false;
    };
    if name != "__init__" {
        return false;
    }
    let Some(parent) = semantic.first_non_type_parent_scope(scope) else {
        return false;
    };

    if !matches!(
        function_type::classify(
            name,
            decorator_list,
            parent,
            semantic,
            &settings.pep8_naming.classmethod_decorators,
            &settings.pep8_naming.staticmethod_decorators,
        ),
        function_type::FunctionType::Method
    ) {
        return false;
    }
    true
}

/// A wrapper around [`CmpOp`] that implements `Display`.
#[derive(Debug)]
pub(super) struct CmpOpExt(CmpOp);

impl From<&CmpOp> for CmpOpExt {
    fn from(cmp_op: &CmpOp) -> Self {
        CmpOpExt(*cmp_op)
    }
}

impl fmt::Display for CmpOpExt {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let representation = match self.0 {
            CmpOp::Eq => "==",
            CmpOp::NotEq => "!=",
            CmpOp::Lt => "<",
            CmpOp::LtE => "<=",
            CmpOp::Gt => ">",
            CmpOp::GtE => ">=",
            CmpOp::Is => "is",
            CmpOp::IsNot => "is not",
            CmpOp::In => "in",
            CmpOp::NotIn => "not in",
        };
        write!(f, "{representation}")
    }
}
