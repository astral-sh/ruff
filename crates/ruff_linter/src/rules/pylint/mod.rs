//! Rules from [Pylint](https://pypi.org/project/pylint/).
mod helpers;
pub(crate) mod rules;
pub mod settings;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use regex::Regex;
    use test_case::test_case;

    use crate::assert_messages;
    use crate::registry::Rule;
    use crate::rules::pylint;
    use crate::settings::types::PythonVersion;
    use crate::settings::LinterSettings;
    use crate::test::test_path;

    #[test_case(Rule::AndOrTernary, Path::new("and_or_ternary.py"))]
    #[test_case(Rule::AssertOnStringLiteral, Path::new("assert_on_string_literal.py"))]
    #[test_case(Rule::AwaitOutsideAsync, Path::new("await_outside_async.py"))]
    #[test_case(
        Rule::BadStringFormatCharacter,
        Path::new("bad_string_format_character.py")
    )]
    #[test_case(Rule::BadStrStripCall, Path::new("bad_str_strip_call.py"))]
    #[test_case(Rule::BadStringFormatType, Path::new("bad_string_format_type.py"))]
    #[test_case(Rule::BidirectionalUnicode, Path::new("bidirectional_unicode.py"))]
    #[test_case(Rule::BinaryOpException, Path::new("binary_op_exception.py"))]
    #[test_case(Rule::CollapsibleElseIf, Path::new("collapsible_else_if.py"))]
    #[test_case(Rule::CompareToEmptyString, Path::new("compare_to_empty_string.py"))]
    #[test_case(Rule::ComparisonOfConstant, Path::new("comparison_of_constant.py"))]
    #[test_case(
        Rule::RepeatedIsinstanceCalls,
        Path::new("repeated_isinstance_calls.py")
    )]
    #[test_case(Rule::ComparisonWithItself, Path::new("comparison_with_itself.py"))]
    #[test_case(Rule::EqWithoutHash, Path::new("eq_without_hash.py"))]
    #[test_case(Rule::ManualFromImport, Path::new("import_aliasing.py"))]
    #[test_case(Rule::SingleStringSlots, Path::new("single_string_slots.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_0.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_1.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_2.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_3.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_4.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_5.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_6.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_7.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_8.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_9.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_10.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_11.py"))]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_12.py"))]
    #[test_case(Rule::ContinueInFinally, Path::new("continue_in_finally.py"))]
    #[test_case(Rule::GlobalStatement, Path::new("global_statement.py"))]
    #[test_case(
        Rule::GlobalVariableNotAssigned,
        Path::new("global_variable_not_assigned.py")
    )]
    #[test_case(Rule::ImportSelf, Path::new("import_self/module.py"))]
    #[test_case(Rule::InvalidAllFormat, Path::new("invalid_all_format.py"))]
    #[test_case(Rule::InvalidAllObject, Path::new("invalid_all_object.py"))]
    #[test_case(Rule::InvalidStrReturnType, Path::new("invalid_return_type_str.py"))]
    #[test_case(Rule::DuplicateBases, Path::new("duplicate_bases.py"))]
    #[test_case(Rule::InvalidCharacterBackspace, Path::new("invalid_characters.py"))]
    #[test_case(Rule::InvalidCharacterEsc, Path::new("invalid_characters.py"))]
    #[test_case(Rule::InvalidCharacterNul, Path::new("invalid_characters.py"))]
    #[test_case(Rule::InvalidCharacterSub, Path::new("invalid_characters.py"))]
    #[test_case(
        Rule::InvalidCharacterZeroWidthSpace,
        Path::new("invalid_characters.py")
    )]
    #[test_case(Rule::InvalidEnvvarDefault, Path::new("invalid_envvar_default.py"))]
    #[test_case(Rule::InvalidEnvvarValue, Path::new("invalid_envvar_value.py"))]
    #[test_case(Rule::IterationOverSet, Path::new("iteration_over_set.py"))]
    #[test_case(Rule::LoggingTooFewArgs, Path::new("logging_too_few_args.py"))]
    #[test_case(Rule::LoggingTooManyArgs, Path::new("logging_too_many_args.py"))]
    #[test_case(Rule::MagicValueComparison, Path::new("magic_value_comparison.py"))]
    #[test_case(
        Rule::NamedExprWithoutContext,
        Path::new("named_expr_without_context.py")
    )]
    #[test_case(Rule::NonlocalWithoutBinding, Path::new("nonlocal_without_binding.py"))]
    #[test_case(Rule::PropertyWithParameters, Path::new("property_with_parameters.py"))]
    #[test_case(Rule::RedefinedLoopName, Path::new("redefined_loop_name.py"))]
    #[test_case(Rule::ReturnInInit, Path::new("return_in_init.py"))]
    #[test_case(Rule::TooManyArguments, Path::new("too_many_arguments.py"))]
    #[test_case(Rule::TooManyBranches, Path::new("too_many_branches.py"))]
    #[test_case(
        Rule::TooManyReturnStatements,
        Path::new("too_many_return_statements.py")
    )]
    #[test_case(Rule::TooManyStatements, Path::new("too_many_statements.py"))]
    #[test_case(Rule::TypeBivariance, Path::new("type_bivariance.py"))]
    #[test_case(
        Rule::TypeNameIncorrectVariance,
        Path::new("type_name_incorrect_variance.py")
    )]
    #[test_case(Rule::TypeParamNameMismatch, Path::new("type_param_name_mismatch.py"))]
    #[test_case(
        Rule::UnexpectedSpecialMethodSignature,
        Path::new("unexpected_special_method_signature.py")
    )]
    #[test_case(
        Rule::UnnecessaryDirectLambdaCall,
        Path::new("unnecessary_direct_lambda_call.py")
    )]
    #[test_case(
        Rule::LoadBeforeGlobalDeclaration,
        Path::new("load_before_global_declaration.py")
    )]
    #[test_case(Rule::UselessElseOnLoop, Path::new("useless_else_on_loop.py"))]
    #[test_case(Rule::UselessImportAlias, Path::new("import_aliasing.py"))]
    #[test_case(Rule::UselessReturn, Path::new("useless_return.py"))]
    #[test_case(
        Rule::YieldFromInAsyncFunction,
        Path::new("yield_from_in_async_function.py")
    )]
    #[test_case(Rule::YieldInInit, Path::new("yield_in_init.py"))]
    #[test_case(Rule::NestedMinMax, Path::new("nested_min_max.py"))]
    #[test_case(
        Rule::RepeatedEqualityComparison,
        Path::new("repeated_equality_comparison.py")
    )]
    #[test_case(Rule::SelfAssigningVariable, Path::new("self_assigning_variable.py"))]
    #[test_case(
        Rule::SubprocessPopenPreexecFn,
        Path::new("subprocess_popen_preexec_fn.py")
    )]
    #[test_case(
        Rule::SubprocessRunWithoutCheck,
        Path::new("subprocess_run_without_check.py")
    )]
    #[test_case(Rule::BadDunderMethodName, Path::new("bad_dunder_method_name.py"))]
    #[test_case(Rule::NoSelfUse, Path::new("no_self_use.py"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.noqa_code(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("pylint").join(path).as_path(),
            &LinterSettings::for_rule(rule_code),
        )?;
        assert_messages!(snapshot, diagnostics);
        Ok(())
    }

    #[test]
    fn repeated_isinstance_calls() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/repeated_isinstance_calls.py"),
            &LinterSettings::for_rule(Rule::RepeatedIsinstanceCalls)
                .with_target_version(PythonVersion::Py39),
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn continue_in_finally() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/continue_in_finally.py"),
            &LinterSettings::for_rule(Rule::ContinueInFinally)
                .with_target_version(PythonVersion::Py37),
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn allow_magic_value_types() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/magic_value_comparison.py"),
            &LinterSettings {
                pylint: pylint::settings::Settings {
                    allow_magic_value_types: vec![pylint::settings::ConstantType::Int],
                    ..pylint::settings::Settings::default()
                },
                ..LinterSettings::for_rule(Rule::MagicValueComparison)
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_args() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_arguments_params.py"),
            &LinterSettings {
                pylint: pylint::settings::Settings {
                    max_args: 4,
                    ..pylint::settings::Settings::default()
                },
                ..LinterSettings::for_rule(Rule::TooManyArguments)
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_args_with_dummy_variables() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_arguments_params.py"),
            &LinterSettings {
                dummy_variable_rgx: Regex::new(r"skip_.*").unwrap(),
                ..LinterSettings::for_rule(Rule::TooManyArguments)
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_branches() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_branches_params.py"),
            &LinterSettings {
                pylint: pylint::settings::Settings {
                    max_branches: 1,
                    ..pylint::settings::Settings::default()
                },
                ..LinterSettings::for_rule(Rule::TooManyBranches)
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_statements() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_statements_params.py"),
            &LinterSettings {
                pylint: pylint::settings::Settings {
                    max_statements: 1,
                    ..pylint::settings::Settings::default()
                },
                ..LinterSettings::for_rule(Rule::TooManyStatements)
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_return_statements() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_return_statements_params.py"),
            &LinterSettings {
                pylint: pylint::settings::Settings {
                    max_returns: 1,
                    ..pylint::settings::Settings::default()
                },
                ..LinterSettings::for_rule(Rule::TooManyReturnStatements)
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn too_many_public_methods() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_public_methods.py"),
            &LinterSettings {
                pylint: pylint::settings::Settings {
                    max_public_methods: 7,
                    ..pylint::settings::Settings::default()
                },
                ..LinterSettings::for_rules(vec![Rule::TooManyPublicMethods])
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }
}
