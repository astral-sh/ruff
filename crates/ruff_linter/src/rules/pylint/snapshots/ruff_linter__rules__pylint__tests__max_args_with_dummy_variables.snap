---
source: crates/ruff_linter/src/rules/pylint/mod.rs
---
too_many_arguments_params.py:9:5: PLR0913 Too many arguments in function definition (6 > 5)
   |
 7 | # Too many args (6/4) for max_args=4
 8 | # Too many args (6/5) for dummy_variable_rgx ~ "skip_.*"
 9 | def f(x, y, z, t, u, v):
   |     ^ PLR0913
10 |     pass
   |


